### PR TITLE
[Merged by Bors] - feat(Algebra/Algebra/Opposite): tools for working with opposite algebras

### PR DESCRIPTION
**Pull request by eric-wieser**
This moves the `Algebra` instance on `MulOpposite` to a new file, and adds the `AlgHom` and `AlgEquiv` versions of various existing `RingHom` and `RingEquiv` definitions:

* `AlgHom.fromOpposite`
* `AlgHom.toOpposite`
* `AlgHom.op`
* `AlgHom.unop`
* `AlgEquiv.op`
* `AlgHom.unop`
* `AlgEquiv.toOpposite`

As `MulOpposite.instAlgebra` is no longer in `Mathlib.Algebra.Algebra.Basic`, some new downstream imports are needed.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Matrix/Basic.lean [hunk 1]
```diff
@@ -3,6 +3,7 @@ Copyright (c) 2018 Ellen Arlt. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Ellen Arlt, Blair Shi, Sean Leather, Mario Carneiro, Johan Commelin, Lu-Ming Zhang
 -/
+import Mathlib.Algebra.Algebra.Opposite
```
Comment (ocfnash) [inline Mathlib/Data/Matrix/Basic.lean:6]: Is this necessary / deliberate?

Comment (ocfnash) [reply Mathlib/Data/Matrix/Basic.lean:6]: Same question for `Mathlib/Algebra/Algebra/Operations.lean`

File: Mathlib/Algebra/Algebra/Opposite.lean [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+import Mathlib.Algebra.Module.Opposites
+import Mathlib.Algebra.Ring.Opposite
+
+/-!
+# Algebra structures on the multiplicative opposite
+
+## Main definitions
+
+* `MulOpposite.instAlgebra`: the algebra on `Aᵐᵒᵖ`
+* `AlgHom.op`/`AlgHom.unop`: simulatenously convert the domain and codomain of a morphism to the
```
Comment (ocfnash) [inline Mathlib/Algebra/Algebra/Opposite.lean:16]:
```suggestion
* `AlgHom.op`/`AlgHom.unop`: simultaneously convert the domain and codomain of a morphism to the
```

File: Mathlib/Algebra/Algebra/Opposite.lean [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+  opposite algebra.
+* `AlgEquiv.op`/`AlgEquiv.unop`: simulatenously convert the source and target of an isomorphism to
```
Comment (ocfnash) [inline Mathlib/Algebra/Algebra/Opposite.lean:18]:
```suggestion
* `AlgEquiv.op`/`AlgEquiv.unop`: simultaneously convert the source and target of an isomorphism to
```

File: Mathlib/Algebra/Algebra/Opposite.lean [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+theorem toLinearMap_toOpposite (f : A →ₐ[R] B) (hf : ∀ x y, Commute (f x) (f y)) :
+    (f.toOpposite hf : A →ₗ[R] Bᵐᵒᵖ) = (opLinearEquiv R : B ≃ₗ[R] Bᵐᵒᵖ) ∘ₗ f.toLinearMap :=
+  rfl
+
+@[simp]
+theorem toRingHom_toOpposite (f : A →ₐ[R] B) (hf : ∀ x y, Commute (f x) (f y)) :
+    (f.toOpposite hf : A →+* Bᵐᵒᵖ) = (f : A →+* B).toOpposite hf :=
+  rfl
+
+/-- An algebra hom `A →ₐ[R] B` can equivalently be viewed as an algebra hom `αᵐᵒᵖ →ₐ[R] Bᵐᵒᵖ`.
```
Comment (ocfnash) [inline Mathlib/Algebra/Algebra/Opposite.lean:94]:
```suggestion
/-- An algebra hom `A →ₐ[R] B` can equivalently be viewed as an algebra hom `Aᵐᵒᵖ →ₐ[R] Bᵐᵒᵖ`.
```

File: Mathlib/Algebra/Algebra/Opposite.lean [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+abbrev unop : (Aᵐᵒᵖ →ₐ[R] Bᵐᵒᵖ) ≃ (A →ₐ[R] B) := AlgHom.op.symm
+
+theorem toRingHom_unop (f : Aᵐᵒᵖ →ₐ[R] Bᵐᵒᵖ) : f.unop.toRingHom = RingHom.unop f.toRingHom :=
+  rfl
+
+end AlgHom
+
+namespace AlgEquiv
+
+/-- An algebra iso `A ≃ₐ[R] B` can equivalently be viewed as an algebra iso `αᵐᵒᵖ ≃ₐ[R] Bᵐᵒᵖ`.
```
Comment (ocfnash) [inline Mathlib/Algebra/Algebra/Opposite.lean:116]:
```suggestion
/-- An algebra iso `A ≃ₐ[R] B` can equivalently be viewed as an algebra iso `Aᵐᵒᵖ ≃ₐ[R] Bᵐᵒᵖ`.
```

**Review by ocfnash — APPROVED** [review]
Thanks!

bors d+

File: Mathlib/Data/Matrix/Basic.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/Data/Matrix/Basic.lean:6]: Yes, the opposite algebra is needed by this file and is no longer provided by `Algebra.Algebra.Basic`. I'll edit the commit message.

**Comment by eric-wieser** [comment]
bors merge

(the changes were only to docstrings)

**Pull request closed without merging**